phd-tests/framework/src/serial/mod.rs

```diff
@@ -2,114 +2,210 @@
 
+//! Interfaces to access a guest's serial console.
+
+use std::sync::{Arc, Mutex};
+
 use anyhow::Result;
+use camino::Utf8PathBuf;
 use futures::{SinkExt, StreamExt};
 use reqwest::Upgraded;
-use tokio::sync::mpsc;
-use tokio::task::JoinHandle;
-use tokio_tungstenite::tungstenite::protocol::Role;
-use tokio_tungstenite::tungstenite::Message;
-use tokio_tungstenite::WebSocketStream;
-use tracing::{error, info, info_span, Instrument};
+use tokio::{
+    sync::mpsc::{UnboundedReceiver, UnboundedSender},
+    task::JoinHandle,
+};
+use tokio_tungstenite::{
+    tungstenite::{protocol::Role, Message},
+    WebSocketStream,
+};
+use tracing::{debug, error, info, info_span, Instrument};
+
+mod raw_buffer;
+
+/// Describes a request to wait for a string to appear on the serial console or
+/// in the console's back buffer.
+struct OutputWaiter {
+    /// The string this waiter should wait for.
+    wanted: String,
+
+    /// When the wait is satisfied, send the contents of the buffer prior to
+    /// (and exclusive of) the waited-for string to this channel.
+    preceding_tx: UnboundedSender<String>,
+}
 
-use self::vt100::Vt100Processor;
+/// An interface for objects that handle and buffer characters and commands a
+/// guest writes to its serial console.
+trait Buffer: Send {
+    /// Processes the supplied `bytes` as input to the buffer.
+    fn process_bytes(&mut self, bytes: &[u8]);
 
-mod vt100;
+    /// Registers a new request to wait for a string to appear in the buffer.
+    fn register_wait_for_output(&mut self, waiter: OutputWaiter);
+
+    /// Ensures there is no active request to wait for a string to appear in
+    /// this buffer. Returns the previous active request if there was one.
+    fn cancel_wait_for_output(&mut self) -> Option<OutputWaiter>;
+}
+
+/// The kind of buffering discipline to use for a guest's serial output.
+pub enum BufferKind {
+    /// Assume that the guest will output characters and command bytes (like
+    /// carriage returns and line feeds) "in the raw" without trying to
+    /// implement its own buffering or scrollback.
+    Raw,
+
+    #[allow(dead_code)]
+    /// Assume that the guest believes it is sending commands to drive a
+    /// VT100-compatible 80x24 terminal and emulate that terminal.
+    Vt80x24,
+}
 
+/// A connection to a guest serial console made available on a particular guest
+/// serial port.
 pub struct SerialConsole {
+    /// A handle to a tokio task that handles websocket messages to and from the
+    /// Propolis server that serves this console.
     ws_task: JoinHandle<()>,
-    vt_processor: vt100::Vt100Processor,
-    wstx: tokio::sync::mpsc::Sender<Vec<u8>>,
+
+    /// Bytes sent to this channel are received by the websocket task, which
+    /// sends them to Propolis.
+    ws_tx: UnboundedSender<Vec<u8>>,
+
+    /// The buffer for this serial console, shared by the console itself and by
+    /// its websocket handler task.
+    buffer: Arc<Mutex<Box<dyn Buffer>>>,
```

Comment:
AFAICT, the buffer only needs to be shared between the `SerialConsole` struct and the websocket task so that we can mutate the `OutputWaiter` in `register_wait_for_string`/`cancel_wait_for_string`. The actual buffer is never touched by the `SerialConsole` struct, just the `OutputWaiter`. It seems like we _could_ refactor this so that we just share an `Arc<Mutex<OutputWaiter>>` between the `SerialConsole` and the websocket task, so that we don't need to also lock the mutex every time we write bytes when there's no output waiter set?

Potentially, we could even go as far as changing this code to give the websocket task exclusive ownership over the `OutputWaiter` as well, and register a new waiter/cancel the waiter by sending messages to the task over a channel. This way, we have no shared state that requires a mutex, which is potentially much more efficient.

This isn't a blocking concern, but it might be worth looking into.

Reply:
This is going to get a little more complicated in the future, because I intend to make it possible to switch buffering disciplines at runtime (to handle guests that e.g. use `RawBuffer` at first but will switch to `Vt80x24` midway through their startup sequences). My thinking is that `SerialConsole` would handle that case by taking the lock and swapping out the existing `Box` with a pointer to a new buffer that implements the new discipline.

I did think a bit about introducing a separate control channel so that the websocket task can just own everything (in this world it would also need to accept messages that say "change your buffering mode to X"). I can give that another whirl and see how it looks now that everything else is more or less put together.

Reply:
I like how this turned out--it also makes it a little easier to fail fast in cases where, e.g., the task exited because something went wrong with the websocket connection (previously, you could register new waits but would have to wait for them to time out; now, to register a wait, you have to send to the task, so you error out right away if the connection was lost). Retested with Alpine and Debian 11 and pushed in 119d57b9.

Reply:
:woman_shrugging: the mutex makes more sense given those future plans, i'm fine with this approach. again, the cost of any of this stuff is almost certainly not what dominates test run time. :)

Reply:
ah, looks like we raced on comments again. i like the approach in i like the approach you switched to in [119d57b](https://github.com/oxidecomputer/propolis/pull/615/commits/119d57b913d11c86e1b2f7a19856d4837abf68d7) a lot, this seems great! in the future, if we want to add code for switching the buffer discipline at runtime, i suppose we could do that over the command channel, too!